sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py

```diff
@@ -423,25 +444,45 @@ def refresh(self, **kwargs) -> None:
-                        headers=_get_headers("Watch", uses_key_vault=self._uses_key_vault, **kwargs),
-                        **kwargs
-                    )
-                    if updated_sentinel is not None:
-                        logging.debug(
-                            "Refresh all triggered by key: %s label %s.",
-                            key,
-                            label,
+                need_refresh = False
+                updated_sentinel_keys = dict(self._refresh_on)
+                headers = _get_headers("Watch")
+                for (key, label), etag in updated_sentinel_keys.items():
+                    try:
+                        updated_sentinel = self._client.get_configuration_setting(
+                            key=key,
+                            label=label,
+                            etag=etag,
+                            match_condition=MatchConditions.IfModified,
+                            headers=headers,
+                            **kwargs
                         )
-                        self._load_all(headers=_get_headers("Watch", uses_key_vault=self._uses_key_vault, **kwargs))
-                        self._refresh_on[(key, label)] = updated_sentinel.etag
-                        self._refresh_timer.reset()
-                        return
+                        if updated_sentinel is not None:
+                            logging.debug(
+                                "Refresh all triggered by key: %s label %s.",
+                                key,
+                                label,
+                            )
+                            need_refresh = True
+
+                            updated_sentinel_keys[(key, label)] = updated_sentinel.etag
+                    except HttpResponseError as e:
+                        if e.status_code == 404:
+                            if etag is not None:
+                                # If the sentinel is not found, it means the key/label was deleted, so we should refresh
+                                logging.debug("Refresh all triggered by key: %s label %s.", key, label)
+                                need_refresh = True
+                                updated_sentinel_keys[(key, label)] = None
+                        else:
+                            raise e
+                # Need to only update once, no matter how many sentinels are updated
+                if need_refresh:
+                    self._load_all(sentinel_keys=updated_sentinel_keys, **kwargs)
+                    if self._on_refresh_success:
+                        self._on_refresh_success()
+                # Even if we don't need to refresh, we should reset the timer
+                self._refresh_timer.reset()
+                return
         except (ServiceRequestError, ServiceResponseError) as e:
             logging.debug("Failed to refresh, retrying: %r", e)
             self._refresh_timer.retry()
```

Comment:
The name `retry()` throws me off. It sounds like we are retrying, but in reality, we are just updating the next refresh time. Can you call it something like `set_next_refresh_time()`?

Reply:
Also, shouldn't we invoke `_on_refresh_error` after this?

Reply:
> Also, shouldn't we invoke `_on_refresh_error` after this?

I think the logic here was that this is a retriable error, therefore. we don't need to raise the error.

Reply:
@avanigupta, how about `backoff()` instead of `retry()` as that is what we are doing.

Reply:
> I think the logic here was that this is a retriable error, therefore. we don't need to raise the error.

While the error may be retriable, we still need clients to make that `refresh` call again in order to 'retry'.  So technically, it's not a retry - it's a new refresh operation. 

`backoff()` also sounds good.

Reply:
I guess my argument would be that there isn't anything in a callback that could be done to trigger another refresh, unless you want to get recursive, which might be way too soon.

Reply:
Maybe we can log it as a retryable error and still failing the refresh operation?
Currently, if the refresh operation ends with a retriable error, we dont call `on_refresh_success()` or `on_refresh_error()`. So the refresh operation neither succeeded nor failed?

Reply:
I think the idea is that whatever this is tied to trigger the refresh happens way more than it actually needs to, just most of the time we return right away before doing anything.

Reply:
> most of the time we return right away before doing anything. 

But this scenario is different. We didn't return right away - we tried to refresh and failed (even after SDK retries). I think users should know that refresh failed so that they can handle it as they see fit.

Reply:
I think the question is what actionable thing could be done. They just need to wait what is probably a few ms and try again.

Reply:
I'm concerned about the scenario where an application is constantly failing for a long time due to retriable errors.  Users might observe that their application is not picking up the config changes even after changing sentinel keys. And they have no idea why that's happening because there are no errors during refresh. 

Failing the refresh in case of such errors is also consistent with .NET provider.